docs/standard/serialization/binaryformatter-migration-guide/reading-nrbf-payloads.md

```diff
@@ -0,0 +1,206 @@
+
+`NrbfDecoder` is following these principles to read from **untrusted input**:
+
+- Treating every input as potentially hostile.
+- No type loading of any kind (to avoid remote code execution).
+- No recursion of any kind (to avoid unbound recursion, stack overflow and denial of service).
+- No buffer pre-allocation based on size provided in the payload, if the payload is too small to contain the promised data (to avoid running out of memory and denial of service).
+- Decoding every part of the input only once (to perform the same amount of work as the potential attacker who created the payload).
+- Using collision-resistant randomized hashing to store records referenced by other records (to avoid running out of memory for dictionary backed by an array which size depends on the number of hash code collisions).
+- Only primitive types can be instantiated in an implicit way. Arrays can be instantiated on demand. Other types are never instantiated.
+
+### Deserializing a closed set of types
+
+`NrbfDecoder` is mostly useful only when the list of serialized types is a known, closed set. Or put it otherwise, you need to know up front what you want to read, because at the end of the day you also need to create instances of these types and populate them with data that was read from the payload. Let's consider two opposite examples.
+
+All `[Serializable]` types from [Quartz.NET](https://github.com/search?q=repo%3Aquartznet%2Fquartznet+%5BSerializable%5D+language%3AC%23&type=code&l=C%23) that can be persisted by the library itself are `sealed`, so there are no custom types that the users could create and the payload can contain only known types. They also provide public constructors, so it is possible to re-create these types based on the information read from payload.
+
+`SettingsPropertyValue` type from `System.Configuration.ConfigurationManager` library exposes an `object PropertyValue` that may internally use `BinaryFormatter` to serialize and deserialize any object that was stored in the configuration file. It could be used to store an integer, a custom type, a dictionary or literally anything.  Because of that, **it is impossible to migrate this library without introducing breaking changes to the API, as the payload can contain anything**.
+
+### Identifying BinaryFormatter payload
+
+`NrbfDecoder` provides two `StartsWithPayloadHeader` methods that allow to **check whether given stream or buffer starts with NRBF header**.
+
+Using these methods is recommended for the period of migrating payloads persisted with `BinaryFormatter` to a [different serializer](./choosing-a-serializer.md):
+
+- Check if the payload read from storage is an [NRBF](/openspecs/windows_protocols/ms-nrbf/75b9fe09-be15-475f-85b8-ae7b7558cfe5) payload.
+- If so, read it with `NrbfDecoder`, serialize it back with a new serializer and overwrite the data in the storage.
+- If not, use the new serializer to deserialize the data.
+
```

Comment:
I really expected sample code here, like

```C#
internal static T LoadFromFile<T>(string path)
{
    bool update = false;
    T value;

    using (FileStream stream = File.OpenRead(path))
    {
        if (NrbfDecoder.StartsWithPayloadHeader(stream))
        {
            value = LoadLegacyValue<T>(stream);
            update = true;
        }
        else
        {
            value = LoadNewValue<T>(stream);
        }
    }

    if (update)
    {
        File.WriteAllBytes(path, NewSerializer(value));
    }

    return value;
}
```

Reply:
I had some, but I did not like it and I deleted it. Yours look great!